src/dataflow-types/src/errors.rs

```diff
@@ -10,36 +10,58 @@
 use std::fmt::Display;
 
 use bytes::BufMut;
-use mz_expr::{EvalError, SourceInstanceId};
+use prost::Message;
+use serde::{Deserialize, Serialize};
+
+use mz_expr::EvalError;
 use mz_persist_types::Codec;
+use mz_repr::proto::{TryFromProtoError, TryIntoIfSome};
+use mz_repr::GlobalId;
 
-use serde::{Deserialize, Serialize};
+include!(concat!(env!("OUT_DIR"), "/mz_dataflow_types.errors.rs"));
 
 #[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash)]
 pub enum DecodeError {
     Text(String),
 }
 
-// We only want to support persisting DecodeError for now, and not the full DataflowError, which is
-// a bit more complex.
+impl From<&DecodeError> for ProtoDecodeError {
+    fn from(error: &DecodeError) -> Self {
+        ProtoDecodeError {
+            kind: Some(match error {
+                DecodeError::Text(v) => proto_decode_error::Kind::Text(v.clone()),
```

Comment:
this extra cloning is unfortunate :(, especially because we are basically doing this to immediately call `encode`. do you know why the pattern in mz is that we do this direction through reference and the other direction through and owned value, other than thats just how `Codec` sets it up? I feel like it would be nice to have `encode` take  `self`, and have a helper like:
```
encode_shared(&self, &mut Buf) -> ... 
where Self: Clone
{
    encode(self.clone())
}
```
which will be the same exact number of clones for existing callsites, but allow owned paths to reduce clones

Reply:
I think you're right that encode should take a self receiver. The fact that it takes a `&self` receiver had let to a proliferation of these From<&T> implementations that would be much better as an owned variant. If you have spare cycles I'd love to see that PR

Reply:
I will file an issue!